### PR TITLE
[Fix] Fix the dimension and yaw hack of FCOS3D on nuScenes

### PR DESCRIPTION
**Pull request by Tai-Wang**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Please describe the motivation of this PR and the goal you want to achieve through this PR.
Due to inconsistent definition of nuscbox and our cambox, the transformation of dimension and yaw is required in the pre-processing and post-processing.

## Modification
Please briefly describe what modification is made in this PR.
1. Dimension and yaw transformed to our cambox in the pre-processing.
2. Corresponding transformation in the post-processing functions.
3. Because the yaw angle defined in our cambox is the global yaw instead of local yaw, we move the transformation in the kitti/nuscenes mono dataset into get_target_3d of fcos3d head.
4. Fix the minor errors caused by the change of cam_intrinsic when applying RandomFlip3D.
5. Remove the hack in the mono3d visualization code.

## BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.
Json files of nuScenes and KITTI datasets need to be regenerated.

## Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/datasets/pipelines/transforms_3d.py [hunk 1]
```diff
@@ -119,9 +119,11 @@ def random_flip_data_3d(self, input_dict, direction='horizontal'):
         if 'centers2d' in input_dict:
             assert self.sync_2d is True and direction == 'horizontal', \
                 'Only support sync_2d=True and horizontal flip with images'
-            w = input_dict['img_shape'][1]
+            w = input_dict['ori_shape'][1]
             input_dict['centers2d'][..., 0] = \
                 w - input_dict['centers2d'][..., 0]
+            input_dict['cam_intrinsic'][0][2] = \
+                w - input_dict['cam_intrinsic'][0][2]
```
Comment (Wuziyi616) [inline mmdet3d/datasets/pipelines/transforms_3d.py:126]: Can we add a comment to explain this? This is a bit hard to understand at first glance for beginners.

**Review by Wuziyi616 — COMMENTED** [review]
Good job! Can we verify the correctness by doing some visualizations? It would be great if you can post results in demo, browse_dataset and visualization while doing model inference below.

**Comment by Wuziyi616** [comment]
I think we also need to update FCOS3D page with new checkpoints, and modify the compatibility docs.

**Comment by Tai-Wang** [comment]
> I think we also need to update FCOS3D page with new checkpoints, and modify the compatibility docs.

Yes, that's why I add a label WIP :)

**Comment by Tai-Wang** [comment]
An example to illustrate the correction of camera intrinsic matrix in the RandomFlip3D.
The original image and the corresponding 3D bounding box:
image
Our image flip is typically based on the horizontal center (e.g. a 1600x900 image is flipped around u=800 pixels). However, the horizontal position of camera center in the image is typically not strictly equal to 800 (for example, sometimes 780 or 820), so it will yield errors when flipping images without manipulating the camera intrinsic matrix accordingly.
Without modifying the camera intrinsics, the visualization will be:
image
With our correction, it will be:
image
So we can see the necessity of modifying the camera intrinsics and have an intuitive feeling of the yielded error from this example.

File: mmdet3d/datasets/pipelines/transforms_3d.py (hunk 1 above)
Comment (Tai-Wang) [reply mmdet3d/datasets/pipelines/transforms_3d.py:131]: Done.

**Review by THU17cyz — APPROVED** [review]
Awesome! After this pr is merged I will check on visualization on the coordinate system refactor branch.

**Comment by THU17cyz** [comment]
> Awesome! After this pr is merged I will check on visualization on the coordinate system refactor branch.

Oh sorry, I should not have approved since there is still work going on.

**Comment by Isaac-xie** [comment]
@Tai-Wang Hi, I'm confused about this modification. In transfroms_3d.py I don't see any changes about any parameters, the parameters that were changed are only centers_2d and cam2img.
First, the ux of cam2img is changed, but the change of  ux only causes the image to be panned and not flipped.
Secondly, after the image is flipped, the yaw angle is changed, doesn't this parameter need to be modified?

**Comment by Isaac-xie** [comment]
Hi, I'm confused about this modification. In transfroms_3d.py I only see changes about centers_2d and cam2img.
after the image is flipped, the yaw angle is changed, doesn't this parameter need to be modified?

**Pull request merged**